File: app/src/main/java/com/example/jonayskieprints/fragment/AdminHomeFragment.java
package com.example.jonayskieprints.fragment;

import android.graphics.Color;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.jonayskieprints.R;
import com.example.jonayskieprints.activity.AdminDashboardActivity;
import com.example.jonayskieprints.adapter.OrderAdapter;
import com.example.jonayskieprints.api.UserApiClient;
import com.example.jonayskieprints.model.Order;
import com.github.mikephil.charting.charts.PieChart;
import com.github.mikephil.charting.data.PieData;
import com.github.mikephil.charting.data.PieDataSet;
import com.github.mikephil.charting.data.PieEntry;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

public class AdminHomeFragment extends Fragment {

    private static final int POLL_INTERVAL = 15_000;

    private TextView     tvLastUpdated;
    private PieChart     pieChart;
    private RecyclerView rvRecentOrders;
    private TextView     tvNoOrders;

    private TextView tvTotalOrders, tvPendingOrders, tvTotalCustomers, tvTotalSales;
    private TextView tvLabelOrders, tvLabelPending, tvLabelCustomers, tvLabelSales;

    private OrderAdapter recentAdapter;

    private final Handler  handler     = new Handler(Looper.getMainLooper());
    private final Runnable pollRunnable = this::fetchData;

    @Nullable @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_admin_home, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        tvLastUpdated  = view.findViewById(R.id.tv_last_updated);
        pieChart       = view.findViewById(R.id.pie_chart);
        rvRecentOrders = view.findViewById(R.id.rv_recent_orders);
        tvNoOrders     = view.findViewById(R.id.tv_no_orders);

        // ── Stat cards ────────────────────────────────────────────────────────
        View cardOrders    = view.findViewById(R.id.layout_total_orders);
        View cardPending   = view.findViewById(R.id.layout_pending_orders);
        View cardCustomers = view.findViewById(R.id.layout_total_customers);
        View cardSales     = view.findViewById(R.id.layout_total_sales);

        tvTotalOrders    = cardOrders.findViewById(R.id.tv_stat_value);
        tvPendingOrders  = cardPending.findViewById(R.id.tv_stat_value);
        tvTotalCustomers = cardCustomers.findViewById(R.id.tv_stat_value);
        tvTotalSales     = cardSales.findViewById(R.id.tv_stat_value);

        tvLabelOrders    = cardOrders.findViewById(R.id.tv_stat_label);
        tvLabelPending   = cardPending.findViewById(R.id.tv_stat_label);
        tvLabelCustomers = cardCustomers.findViewById(R.id.tv_stat_label);
        tvLabelSales     = cardSales.findViewById(R.id.tv_stat_label);

        tvLabelOrders.setText("Total Orders");
        tvLabelPending.setText("Pending");
        tvLabelCustomers.setText("Customers");
        tvLabelSales.setText("Total Sales");

        setIcon(cardOrders,    R.drawable.ic_cart);
        setIcon(cardPending,   R.drawable.ic_clock);
        setIcon(cardCustomers, R.drawable.ic_users);
        setIcon(cardSales,     R.drawable.ic_credit_card);

        // ── Clickable stat cards ──────────────────────────────────────────────
        cardOrders.setOnClickListener(v    -> goToOrders(""));
        cardPending.setOnClickListener(v   -> goToOrders("pending"));
        cardSales.setOnClickListener(v     -> goToOrders("completed"));
        cardCustomers.setOnClickListener(v ->
                requireActivity().getSupportFragmentManager()
                        .beginTransaction()
                        .replace(R.id.fragment_container, new CustomersFragment())
                        .addToBackStack(null)
                        .commit());

        // ── View All button ───────────────────────────────────────────────────
        view.findViewById(R.id.btn_view_all_orders).setOnClickListener(v -> goToOrders(""));

        // ── Recent orders RecyclerView ────────────────────────────────────────
        recentAdapter = new OrderAdapter(
                new ArrayList<>(),
                new OrderAdapter.Listener() {
                    @Override public void onViewDetails(Order o) { /* TODO */ }
                    @Override public void onEdit(Order o) {}
                    @Override public void onCancel(Order o) {}
                },
                false,  // showActions
                false   // isAdminMode — recent list is display-only
        );
        rvRecentOrders.setLayoutManager(new LinearLayoutManager(requireContext()));
        rvRecentOrders.setNestedScrollingEnabled(false);
        rvRecentOrders.setAdapter(recentAdapter);

        setupPieChart();
        fetchData();
        handler.postDelayed(pollRunnable, POLL_INTERVAL);
    }

    private void setIcon(View card, int drawableRes) {
        android.widget.ImageView iv = card.findViewById(R.id.iv_stat_icon);
        if (iv != null) {
            iv.setImageResource(drawableRes);
            iv.setColorFilter(Color.WHITE, android.graphics.PorterDuff.Mode.SRC_IN);
        }
    }

    private void setupPieChart() {
        pieChart.getDescription().setEnabled(false);
        pieChart.setDrawHoleEnabled(true);
        pieChart.setHoleRadius(55f);
        pieChart.setTransparentCircleRadius(60f);
        pieChart.setHoleColor(Color.WHITE);
        pieChart.setRotationEnabled(true);
        pieChart.getLegend().setEnabled(false);
        pieChart.setCenterTextSize(14f);
        pieChart.setCenterTextColor(Color.parseColor("#111827"));
    }

    // ── Fetch all data ────────────────────────────────────────────────────────
    private void fetchData() {
        if (!isAdded()) return;

        // 1. Stats API → totalOrders, pendingOrders, totalSpent
        UserApiClient.get().fetchStats(stats -> {
            if (!isAdded()) return;

            tvTotalOrders.setText(String.valueOf(stats.totalOrders));
            tvPendingOrders.setText(String.valueOf(stats.pendingOrders));
            tvTotalSales.setText(String.format("₱%.2f", stats.totalSpent));

            String time = new SimpleDateFormat("hh:mm:ss a", Locale.getDefault())
                    .format(new Date());
            tvLastUpdated.setText("Updated " + time);

            if (requireActivity() instanceof AdminDashboardActivity) {
                ((AdminDashboardActivity) requireActivity())
                        .updateNotificationBadge(stats.pendingOrders);
            }
        });

        // 2. All orders → pie chart + unique customer count + recent list
        UserApiClient.get().fetchOrders("", orders -> {
            if (!isAdded()) return;

            int pending = 0, inProgress = 0, completed = 0, cancelled = 0;
            Set<String> uniqueCustomers = new HashSet<>();

            for (Order o : orders) {
                String s = o.status != null ? o.status.toLowerCase() : "";
                switch (s) {
                    case "pending":     pending++;    break;
                    case "in progress":
                    case "inprogress":
                    case "in-progress":
                    case "printing":    inProgress++; break;
                    case "completed":   completed++;  break;
                    case "cancelled":   cancelled++;  break;
                }

                // user_id is stored in userName field (mapped in parseOrder)
                if (o.userName != null && !o.userName.isEmpty()) {
                    uniqueCustomers.add(o.userName);
                }
            }

            // Customers card
            tvTotalCustomers.setText(String.valueOf(uniqueCustomers.size()));

            // Pie chart
            updatePieChart(pending, inProgress, completed, cancelled, orders.size());

            // Recent orders (top 5)
            List<Order> recent = orders.size() > 5
                    ? orders.subList(0, 5) : orders;

            if (recent.isEmpty()) {
                tvNoOrders.setVisibility(View.VISIBLE);
                rvRecentOrders.setVisibility(View.GONE);
            } else {
                tvNoOrders.setVisibility(View.GONE);
                rvRecentOrders.setVisibility(View.VISIBLE);
                recentAdapter.updateOrders(new ArrayList<>(recent));
            }
        });
    }

    // ── Pie chart ─────────────────────────────────────────────────────────────
    private void updatePieChart(int pending, int inProgress,
                                int completed, int cancelled, int total) {
        if (total == 0) {
            pieChart.clear();
            pieChart.setCenterText("No Data");
            pieChart.invalidate();
            return;
        }

        List<PieEntry> entries = new ArrayList<>();
        if (pending    > 0) entries.add(new PieEntry(pending,    "Pending"));
        if (inProgress > 0) entries.add(new PieEntry(inProgress, "In Progress"));
        if (completed  > 0) entries.add(new PieEntry(completed,  "Completed"));
        if (cancelled  > 0) entries.add(new PieEntry(cancelled,  "Cancelled"));

        PieDataSet dataSet = new PieDataSet(entries, "");
        dataSet.setColors(
                Color.parseColor("#F59E0B"),  // pending     — amber
                Color.parseColor("#3B82F6"),  // in progress — blue
                Color.parseColor("#22C55E"),  // completed   — green
                Color.parseColor("#EF4444")   // cancelled   — red
        );
        dataSet.setDrawValues(false);
        dataSet.setSliceSpace(2f);

        pieChart.setData(new PieData(dataSet));
        pieChart.setCenterText(total + "\nTotal Orders");
        pieChart.animateY(800);
        pieChart.invalidate();
    }

    // ── Navigation helper ────────────────────────────────────────────────────
    private void goToOrders(String filter) {
        Bundle args = new Bundle();
        args.putString("filter", filter);
        OrdersFragment frag = new OrdersFragment();
        frag.setArguments(args);
        requireActivity().getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.fragment_container, frag)
                .addToBackStack(null)
                .commit();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        handler.removeCallbacks(pollRunnable);
    }
}

File: app/src/main/java/com/example/jonayskieprints/AdminHomeFragment.kt
package com.example.jonayskieprints

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.example.jonayskieprints.databinding.FragmentAdminHomeBinding
import com.github.mikephil.charting.data.PieData
import com.github.mikephil.charting.data.PieDataSet
import com.github.mikephil.charting.data.PieEntry
import com.github.mikephil.charting.formatter.ValueFormatter
import com.google.firebase.firestore.FirebaseFirestore
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class AdminHomeFragment : Fragment() {

    private var _binding: FragmentAdminHomeBinding? = null
    private val binding get() = _binding!!
    private val db = FirebaseFirestore.getInstance()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        Log.d("AdminHome", "onCreateView called")
        _binding = FragmentAdminHomeBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        Log.d("AdminHome", "onViewCreated called")

        // ── Labels ────────────────────────────────────────────
        binding.layoutTotalOrders.tvStatLabel.text    = "Total Orders"
        binding.layoutPendingOrders.tvStatLabel.text  = "Pending"
        binding.layoutTotalCustomers.tvStatLabel.text = "Customers"
        binding.layoutTotalSales.tvStatLabel.text     = "Total Sales"

        // ── Icons ─────────────────────────────────────────────
        binding.layoutTotalOrders.ivStatIcon.setImageResource(R.drawable.ic_cart)
        binding.layoutPendingOrders.ivStatIcon.setImageResource(R.drawable.ic_clock)
        binding.layoutTotalCustomers.ivStatIcon.setImageResource(R.drawable.ic_users)
        binding.layoutTotalSales.ivStatIcon.setImageResource(R.drawable.ic_credit_card)

        // ── Tint all icons white explicitly ───────────────────
        val white = android.graphics.Color.WHITE
        binding.layoutTotalOrders.ivStatIcon
            .setColorFilter(white, android.graphics.PorterDuff.Mode.SRC_IN)
        binding.layoutPendingOrders.ivStatIcon
            .setColorFilter(white, android.graphics.PorterDuff.Mode.SRC_IN)
        binding.layoutTotalCustomers.ivStatIcon
            .setColorFilter(white, android.graphics.PorterDuff.Mode.SRC_IN)
        binding.layoutTotalSales.ivStatIcon
            .setColorFilter(white, android.graphics.PorterDuff.Mode.SRC_IN)

        loadStatsFromFirestore()
    }

    private fun loadStatsFromFirestore() {
        Log.d("AdminHome", "loadStatsFromFirestore() called")

        // ── Fetch all orders ──────────────────────────────────
        db.collection("orders").get()
            .addOnSuccessListener { orders ->
                Log.d("AdminHome", "Orders SUCCESS - count: ${orders.size()}")
                if (_binding == null) {
                    Log.d("AdminHome", "Binding is null, skipping update")
                    return@addOnSuccessListener
                }

                var totalOrders   = 0
                var pending       = 0
                var inProgress    = 0
                var completed     = 0
                var cancelled     = 0
                var totalSales    = 0.0

                for (doc in orders) {
                    totalOrders++
                    val status = doc.getString("status")?.lowercase() ?: ""
                    val amount = doc.getDouble("totalAmount") ?: 0.0
                    Log.d("AdminHome", "Order doc: status='$status', amount=$amount, allFields=${doc.data?.keys}")

                    when (status) {
                        "pending"     -> pending++
                        "in progress",
                        "inprogress"  -> inProgress++
                        "completed"   -> { completed++; totalSales += amount }
                        "cancelled"   -> cancelled++
                    }
                }

                Log.d("AdminHome", "Results => total=$totalOrders, pending=$pending, inProgress=$inProgress, completed=$completed, cancelled=$cancelled, sales=$totalSales")

                // ── Update stat cards ─────────────────────────
                binding.layoutTotalOrders.tvStatValue.text   = totalOrders.toString()
                binding.layoutPendingOrders.tvStatValue.text = pending.toString()
                binding.layoutTotalSales.tvStatValue.text    = "₱%.2f".format(totalSales)

                // ── Update timestamp ──────────────────────────
                val time = SimpleDateFormat("hh:mm:ss a", Locale.getDefault()).format(Date())
                binding.tvLastUpdated.text = "Updated $time"

                // ── Draw pie chart ────────────────────────────
                setupPieChart(pending, inProgress, completed, cancelled, totalOrders)
            }
            .addOnFailureListener { e ->
                Log.e("AdminHome", "Orders FAILED: ${e.message}", e)
                if (_binding == null) return@addOnFailureListener
                binding.layoutTotalOrders.tvStatValue.text    = "—"
                binding.layoutPendingOrders.tvStatValue.text  = "—"
                binding.layoutTotalCustomers.tvStatValue.text = "—"
                binding.layoutTotalSales.tvStatValue.text     = "—"
            }

        // ── Fetch customer count separately ───────────────────
        db.collection("users")
            .whereEqualTo("role", "user")
            .get()
            .addOnSuccessListener { users ->
                Log.d("AdminHome", "Users SUCCESS - count: ${users.size()}")
                if (_binding == null) return@addOnSuccessListener
                binding.layoutTotalCustomers.tvStatValue.text = users.size().toString()
            }
            .addOnFailureListener { e ->
                Log.e("AdminHome", "Users FAILED: ${e.message}", e)
            }
    }

    private fun setupPieChart(
        pending: Int, inProgress: Int,
        completed: Int, cancelled: Int,
        total: Int
    ) {
        if (_binding == null) return

        val chart = binding.pieChart

        if (total == 0) {
            chart.setNoDataText("No chart data available.")
            chart.setNoDataTextColor(android.graphics.Color.parseColor("#F59E0B"))
            chart.invalidate()
            return
        }

        val entries = mutableListOf<PieEntry>()
        if (pending    > 0) entries.add(PieEntry(pending.toFloat(),    "Pending"))
        if (inProgress > 0) entries.add(PieEntry(inProgress.toFloat(), "In Progress"))
        if (completed  > 0) entries.add(PieEntry(completed.toFloat(),  "Completed"))
        if (cancelled  > 0) entries.add(PieEntry(cancelled.toFloat(),  "Cancelled"))

        val colors = listOf(
            android.graphics.Color.parseColor("#F59E0B"),
            android.graphics.Color.parseColor("#3B82F6"),
            android.graphics.Color.parseColor("#22C55E"),
            android.graphics.Color.parseColor("#EF4444")
        )

        val dataSet = PieDataSet(entries, "").apply {
            this.colors = colors
            sliceSpace      = 2f
            selectionShift  = 4f
            setDrawValues(false)
        }

        val data = PieData(dataSet).apply {
            setValueFormatter(object : ValueFormatter() {
                override fun getFormattedValue(value: Float) = "${value.toInt()}"
            })
        }

        chart.apply {
            this.data = data
            isDrawHoleEnabled       = true
            holeRadius              = 58f
            transparentCircleRadius = 62f
            setHoleColor(android.graphics.Color.WHITE)
            description.isEnabled   = false
            legend.isEnabled        = false
            setDrawEntryLabels(false)
            isRotationEnabled       = true
            animateY(800)
            setDrawCenterText(true)
            centerText       = "$total\nTotal Orders"
            setCenterTextSize(14f)
            setCenterTextColor(android.graphics.Color.parseColor("#111827"))
            invalidate()
        }

        updateLegend(pending, inProgress, completed, cancelled)
    }

    private fun updateLegend(
        pending: Int, inProgress: Int,
        completed: Int, cancelled: Int
    ) {
        val legendLayout = binding.llChartLegend
        legendLayout.removeAllViews()

        val items = listOf(
            Triple("#F59E0B", "Pending",     pending),
            Triple("#3B82F6", "In Progress", inProgress),
            Triple("#22C55E", "Completed",   completed),
            Triple("#EF4444", "Cancelled",   cancelled)
        )

        for ((color, label, count) in items) {
            val row = LayoutInflater.from(context)
                .inflate(R.layout.item_chart_legend_dot, legendLayout, false)

            val dot = row.findViewById<View>(R.id.view_legend_dot)
            val bg = android.graphics.drawable.GradientDrawable()
            bg.shape = android.graphics.drawable.GradientDrawable.OVAL
            bg.setColor(android.graphics.Color.parseColor(color))
            dot.background = bg

            row.findViewById<android.widget.TextView>(R.id.tv_legend_label)
                .text = "$label ($count)"

            legendLayout.addView(row)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}